tests/models/test_pyro.py

```diff
@@ -0,0 +1,196 @@
+import os
+
+import numpy as np
+import pyro
+import pyro.distributions as dist
+import torch
+import torch.nn as nn
+from pyro.infer.autoguide import AutoDiagonalNormal
+from pyro.nn import PyroModule, pyro_method
+
+from scvi import _CONSTANTS
+from scvi.compose import DecoderSCVI, Encoder, PyroBaseModuleClass
+from scvi.data import synthetic_iid
+from scvi.dataloaders import AnnDataLoader
+from scvi.lightning import PyroTrainingPlan, Trainer
+
+
+class BayesianRegression(PyroModule, PyroBaseModuleClass):
+    def __init__(self, in_features, out_features):
+        super().__init__()
+
+        self._auto_guide = AutoDiagonalNormal(self)
```

Comment:
Technical detail: One of the rules of `PyroModule`s is that the model and guide must be separate `PyroModule` s, and cannot be contained in a single `PyroModule`; however they can be contained in a single `nn.Module`. You might consider refactoring to separate `PyroBaseModuleClass` from some sort of `PyroModelClass` like this (where `s <: t` means `issubclass(s, t)`):
```
PyroBaseModuleClass <: nn.Module
  .model : PyroBaseModelClass <: PyroModule
  .guide : AutoNormal <: PyroModule
```
Basically "a model can't have its guide as an attribute".

<details>
The reason is due to `PyroModule` naming schemes and caching of `pyro.sample` calls. If `model` and `guide` are contained in a single `PyroModule` there may be weird conflicts in both names and the `pyro.sample` cache.
</details>

Reply:
this is super helpful!